In [30]:
import pandas as pd
from nba_api.stats.static import players, teams
from nba_api.stats.endpoints import (
    leaguedashplayerstats, boxscoretraditionalv3, playercareerstats,
    teamgamelog, drafthistory, leagueleaders,
)

In [5]:
# Fetching all players and creating a DataFrame - Static Data

players_df = pd.DataFrame(players.get_players())
display(players_df.head())
print(f"rows: {players_df.shape[0]}, columns: {players_df.shape[1]}")
players_df.columns.tolist()

,id,full_name,first_name,last_name,is_active
0,76001,Alaa Abdelnaby,Alaa,Abdelnaby,False
1,76002,Zaid Abdul-Aziz,Zaid,Abdul-Aziz,False
2,76003,Kareem Abdul-Jabbar,Kareem,Abdul-Jabbar,False
3,51,Mahmoud Abdul-Rauf,Mahmoud,Abdul-Rauf,False
4,1505,Tariq Abdul-Wahad,Tariq,Abdul-Wahad,False


rows: 5103, columns: 5


['id', 'full_name', 'first_name', 'last_name', 'is_active']

### `players_df` — column reference

- **id** (`int64`) — unique player ID used across nba_api endpoints as `player_id`
- **full_name** (`object`) — player's full name, e.g. "LeBron James"
- **first_name** (`object`) — first name only
- **last_name** (`object`) — last name only
- **is_active** (`bool`) — True if currently on an NBA roster, False if retired/historical

Rows: 5,103. Nulls: 0 across all columns (static reference list, always complete).

In [20]:
#Fetching all teams and creating a DataFrame - Static Data

teams_df = pd.DataFrame(teams.get_teams())
display(teams_df.head(20))
print(f"rows: {teams_df.shape[0]}, columns: {teams_df.shape[1]}")
teams_df.columns.tolist()

,id,full_name,abbreviation,nickname,city,state,year_founded
0,1610612737,Atlanta Hawks,ATL,Hawks,Atlanta,Georgia,1949
1,1610612738,Boston Celtics,BOS,Celtics,Boston,Massachusetts,1946
2,1610612739,Cleveland Cavaliers,CLE,Cavaliers,Cleveland,Ohio,1970
3,1610612740,New Orleans Pelicans,NOP,Pelicans,New Orleans,Louisiana,2002
4,1610612741,Chicago Bulls,CHI,Bulls,Chicago,Illinois,1966
5,1610612742,Dallas Mavericks,DAL,Mavericks,Dallas,Texas,1980
6,1610612743,Denver Nuggets,DEN,Nuggets,Denver,Colorado,1976
7,1610612744,Golden State Warriors,GSW,Warriors,San Francisco,California,1946
8,1610612745,Houston Rockets,HOU,Rockets,Houston,Texas,1967
9,1610612746,Los Angeles Clippers,LAC,Clippers,Los Angeles,California,1970


rows: 30, columns: 7


['id',
 'full_name',
 'abbreviation',
 'nickname',
 'city',
 'state',
 'year_founded']

### `teams_df` — column reference

- **id** (`int64`) — unique team ID used across nba_api endpoints as `team_id`
- **full_name** (`object`) — full team name, e.g. "Atlanta Hawks"
- **abbreviation** (`object`) — 3-letter code, e.g. "ATL"
- **nickname** (`object`) — team name without city, e.g. "Hawks"
- **city** (`object`) — team's home city
- **state** (`object`) — team's home state
- **year_founded** (`int64`) — franchise founding year

Rows: 30. Nulls: 0 across all columns (static reference list, all 30 current franchises).

In [29]:
def profile_dataframe(df, name):
    null_counts = df.isnull().sum()
    summary = pd.DataFrame({
        "column": df.columns,
        "dtype": df.dtypes.values,
        "null_count": null_counts.values,
        "null_pct": (null_counts.values / len(df) * 100).round(2),
    })
    print(f"--- {name} ---")
    print(f"rows: {len(df)}, columns: {len(df.columns)}")
    print(summary.to_string(index=False))
    return summary

In [8]:
league_stats = leaguedashplayerstats.LeagueDashPlayerStats(season="2025-26").get_data_frames()[0]

In [9]:
display(league_stats.head())
print(f"rows: {league_stats.shape[0]}, columns: {league_stats.shape[1]}")

,PLAYER_ID,PLAYER_NAME,NICKNAME,TEAM_ID,TEAM_ABBREVIATION,AGE,GP,W,L,W_PCT,...,BLKA_RANK,PF_RANK,PFD_RANK,PTS_RANK,PLUS_MINUS_RANK,NBA_FANTASY_PTS_RANK,DD2_RANK,TD3_RANK,WNBA_FANTASY_PTS_RANK,TEAM_COUNT
0,1630639,A.J. Lawson,A.J.,1610612761,TOR,25.0,24,12,12,0.500,...,125,142,443,438,290,450,284,38,449,1
1,1631260,AJ Green,AJ,1610612749,MIL,26.0,78,31,47,0.397,...,271,540,229,126,496,175,284,38,139,1
2,1642358,AJ Johnson,AJ,1610612742,DAL,21.0,48,9,39,0.188,...,371,158,384,400,394,414,284,38,413,2
3,203932,Aaron Gordon,Aaron,1610612743,DEN,30.0,36,27,9,0.750,...,271,242,143,197,48,247,114,38,241,1
4,1628988,Aaron Holiday,Aaron,1610612745,HOU,29.0,57,38,19,0.667,...,179,274,323,321,99,354,284,38,348,1


rows: 582, columns: 67


In [10]:
league_stats_summary = profile_dataframe(league_stats, "league_stats")

--- league_stats ---
rows: 582, columns: 67
               column   dtype  null_count  null_pct
            PLAYER_ID   int64           0       0.0
          PLAYER_NAME  object           0       0.0
             NICKNAME  object           0       0.0
              TEAM_ID   int64           0       0.0
    TEAM_ABBREVIATION  object           0       0.0
                  AGE float64           0       0.0
                   GP   int64           0       0.0
                    W   int64           0       0.0
                    L   int64           0       0.0
                W_PCT float64           0       0.0
                  MIN float64           0       0.0
                  FGM   int64           0       0.0
                  FGA   int64           0       0.0
               FG_PCT float64           0       0.0
                 FG3M   int64           0       0.0
                 FG3A   int64           0       0.0
              FG3_PCT float64           0       0.0
                  FT

### `league_stats` — column reference

**Identity columns:**
- **PLAYER_ID** (`int64`) — unique player ID, matches `id` in `players_df`
- **PLAYER_NAME** (`object`) — player's full name
- **NICKNAME** (`object`) — first name / shorthand name used on broadcasts
- **TEAM_ID** (`int64`) — unique team ID, matches `id` in `teams_df`
- **TEAM_ABBREVIATION** (`object`) — 3-letter team code
- **AGE** (`float64`) — player's age as of this season

**Season totals (per player, 2025-26 regular season):**
- **GP** — games played, **W**/**L** — team wins/losses in games this player appeared in, **W_PCT** — win percentage
- **MIN** — total minutes played
- Shooting: **FGM/FGA/FG_PCT** (field goals made/attempted/%), **FG3M/FG3A/FG3_PCT** (3-pointers), **FTM/FTA/FT_PCT** (free throws)
- Rebounding: **OREB** (offensive), **DREB** (defensive), **REB** (total)
- **AST** — assists, **TOV** — turnovers, **STL** — steals, **BLK** — blocks made, **BLKA** — blocks *against* (shots blocked while shooting)
- **PF** — personal fouls committed, **PFD** — personal fouls drawn
- **PTS** — total points, **PLUS_MINUS** — team's point differential while this player was on court
- **NBA_FANTASY_PTS** / **WNBA_FANTASY_PTS** — fantasy scoring totals under each league's scoring formula
- **DD2** — double-doubles, **TD3** — triple-doubles

**Rank columns:** every stat above has a matching `<STAT>_RANK` column (e.g. `PTS_RANK`, `AST_RANK`) — the player's league-wide rank for that stat this season. That's 30 of the 67 columns, all following this same suffix pattern rather than needing individual descriptions.

**TEAM_COUNT** (`int64`) — number of different teams this player appeared for this season (>1 means they were traded).

Rows: (fill in from your output — how many players). Columns: 67. Nulls: 0 across all columns — season-aggregated totals, nothing missing.

In [21]:
career_stats = playercareerstats.PlayerCareerStats(player_id="2544").get_data_frames()[0] # Dont hardcode thw player ID
display(career_stats.head())
print(f"rows: {career_stats.shape[0]}, columns: {career_stats.shape[1]}")

,PLAYER_ID,SEASON_ID,LEAGUE_ID,TEAM_ID,TEAM_ABBREVIATION,PLAYER_AGE,GP,GS,MIN,FGM,...,FT_PCT,OREB,DREB,REB,AST,STL,BLK,TOV,PF,PTS
0,2544,2003-04,00,1610612739,CLE,19.0,79,79,3120,622,...,0.754,99,333,432,465,130,58,273,149,1654
1,2544,2004-05,00,1610612739,CLE,20.0,80,80,3388,795,...,0.750,111,477,588,577,177,52,262,146,2175
2,2544,2005-06,00,1610612739,CLE,21.0,79,79,3361,875,...,0.738,75,481,556,521,123,66,260,181,2478
3,2544,2006-07,00,1610612739,CLE,22.0,78,78,3190,772,...,0.698,83,443,526,470,125,55,250,171,2132
4,2544,2007-08,00,1610612739,CLE,23.0,75,74,3027,794,...,0.712,133,459,592,539,138,81,255,165,2250


rows: 23, columns: 27


In [12]:
career_stats_summary = profile_dataframe(career_stats, "career_stats")

--- career_stats ---
rows: 23, columns: 27
           column   dtype  null_count  null_pct
        PLAYER_ID   int64           0       0.0
        SEASON_ID  object           0       0.0
        LEAGUE_ID  object           0       0.0
          TEAM_ID   int64           0       0.0
TEAM_ABBREVIATION  object           0       0.0
       PLAYER_AGE float64           0       0.0
               GP   int64           0       0.0
               GS   int64           0       0.0
              MIN   int64           0       0.0
              FGM   int64           0       0.0
              FGA   int64           0       0.0
           FG_PCT float64           0       0.0
             FG3M   int64           0       0.0
             FG3A   int64           0       0.0
          FG3_PCT float64           0       0.0
              FTM   int64           0       0.0
              FTA   int64           0       0.0
           FT_PCT float64           0       0.0
             OREB   int64           0       0

### `career_stats` — column reference

- **PLAYER_ID** (`int64`) — unique player ID, matches `id` in `players_df`
- **SEASON_ID** (`object`) — season identifier, e.g. "2023-24"
- **LEAGUE_ID** (`object`) — league code (NBA = "00")
- **TEAM_ID** (`int64`) — team played for that season, matches `id` in `teams_df`
- **TEAM_ABBREVIATION** (`object`) — 3-letter team code for that season
- **PLAYER_AGE** (`float64`) — player's age during that season
- **GP** (`int64`) — games played
- **GS** (`int64`) — games started
- **MIN** (`int64`) — total minutes played
- **FGM** / **FGA** / **FG_PCT** (`int64`/`int64`/`float64`) — field goals made, attempted, percentage
- **FG3M** / **FG3A** / **FG3_PCT** — same, for 3-pointers
- **FTM** / **FTA** / **FT_PCT** — same, for free throws
- **OREB** / **DREB** / **REB** (`int64`) — offensive, defensive, and total rebounds
- **AST** (`int64`) — assists
- **STL** (`int64`) — steals
- **BLK** (`int64`) — blocks
- **TOV** (`int64`) — turnovers
- **PF** (`int64`) — personal fouls
- **PTS** (`int64`) — total points

Rows: 23 (one row per season played). Columns: 27. Nulls: 0 across all columns — season-aggregated totals, same as `league_stats`.

In [25]:
team_gamelog = teamgamelog.TeamGameLog(team_id="1610612747", season="2025-26").get_data_frames()[0]
display(team_gamelog.head())
print(f"rows: {team_gamelog.shape[0]}, columns: {team_gamelog.shape[1]}")

,Team_ID,Game_ID,GAME_DATE,MATCHUP,WL,W,L,W_PCT,MIN,FGM,...,FT_PCT,OREB,DREB,REB,AST,STL,BLK,TOV,PF,PTS
0,1610612747,0022501198,"APR 12, 2026",LAL vs. UTA,W,53,29,0.646,240,51,...,0.824,11,38,49,37,12,6,15,20,131
1,1610612747,0022501185,"APR 10, 2026",LAL vs. PHX,W,52,29,0.642,240,35,...,0.767,4,31,35,27,17,3,11,18,101
2,1610612747,0022501170,"APR 09, 2026",LAL @ GSW,W,51,29,0.638,240,49,...,0.625,8,25,33,37,14,3,17,13,119
3,1610612747,0022501155,"APR 07, 2026",LAL vs. OKC,L,50,29,0.633,240,32,...,0.452,8,30,38,26,7,3,17,12,87
4,1610612747,0022501140,"APR 05, 2026",LAL @ DAL,L,50,28,0.641,240,47,...,0.788,11,38,49,36,6,1,12,23,128


rows: 82, columns: 27


In [14]:
team_gamelog_summary = profile_dataframe(team_gamelog, "team_gamelog")

--- team_gamelog ---
rows: 82, columns: 27
   column   dtype  null_count  null_pct
  Team_ID   int64           0       0.0
  Game_ID  object           0       0.0
GAME_DATE  object           0       0.0
  MATCHUP  object           0       0.0
       WL  object           0       0.0
        W   int64           0       0.0
        L   int64           0       0.0
    W_PCT float64           0       0.0
      MIN   int64           0       0.0
      FGM   int64           0       0.0
      FGA   int64           0       0.0
   FG_PCT float64           0       0.0
     FG3M   int64           0       0.0
     FG3A   int64           0       0.0
  FG3_PCT float64           0       0.0
      FTM   int64           0       0.0
      FTA   int64           0       0.0
   FT_PCT float64           0       0.0
     OREB   int64           0       0.0
     DREB   int64           0       0.0
      REB   int64           0       0.0
      AST   int64           0       0.0
      STL   int64           0       0

### `team_gamelog` — column reference

- **Team_ID** (`int64`) — team ID, matches `id` in `teams_df`
- **Game_ID** (`object`) — unique ID for that specific game, matches `gameId`/`GAME_ID` used in box score endpoints
- **GAME_DATE** (`object`) — date the game was played
- **MATCHUP** (`object`) — opponent and home/away indicator, e.g. "LAL vs. BOS" (home) or "LAL @ BOS" (away)
- **WL** (`object`) — win ("W") or loss ("L") for this team in this game
- **W** / **L** (`int64`) — cumulative win/loss count through this point in the season
- **W_PCT** (`float64`) — cumulative win percentage through this point in the season
- **MIN** (`int64`) — total team minutes played (game + any overtime)
- **FGM/FGA/FG_PCT**, **FG3M/FG3A/FG3_PCT**, **FTM/FTA/FT_PCT** — team shooting totals for that single game (made/attempted/percentage, for field goals, 3-pointers, free throws)
- **OREB/DREB/REB** (`int64`) — offensive, defensive, total team rebounds
- **AST** (`int64`) — team assists, **STL** — steals, **BLK** — blocks, **TOV** — turnovers, **PF** — personal fouls
- **PTS** (`int64`) — team's total points scored that game

Rows: 82 (for each game). Columns: 27. Nulls: 0.

In [15]:
draft_history = drafthistory.DraftHistory().get_data_frames()[0]
display(draft_history.head())
print(f"rows: {draft_history.shape[0]}, columns: {draft_history.shape[1]}")

,PERSON_ID,PLAYER_NAME,SEASON,ROUND_NUMBER,ROUND_PICK,OVERALL_PICK,DRAFT_TYPE,TEAM_ID,TEAM_CITY,TEAM_NAME,TEAM_ABBREVIATION,ORGANIZATION,ORGANIZATION_TYPE,PLAYER_PROFILE_FLAG
0,1643407,AJ Dybantsa,2026,1,1,1,Draft,1610612764,Washington,Wizards,WAS,Brigham Young,College/University,1
1,1643408,Darryn Peterson,2026,1,2,2,Draft,1610612762,Utah,Jazz,UTA,Kansas,College/University,1
2,1643409,Cameron Boozer,2026,1,3,3,Draft,1610612763,Memphis,Grizzlies,MEM,Duke,College/University,1
3,1643410,Caleb Wilson,2026,1,4,4,Draft,1610612741,Chicago,Bulls,CHI,North Carolina,College/University,1
4,1643413,Keaton Wagler,2026,1,5,5,Draft,1610612746,LA,Clippers,LAC,Illinois,College/University,1


rows: 8434, columns: 14


In [16]:
draft_history_summary = profile_dataframe(draft_history, "draft_history")

--- draft_history ---
rows: 8434, columns: 14
             column  dtype  null_count  null_pct
          PERSON_ID  int64           0       0.0
        PLAYER_NAME object           0       0.0
             SEASON object           0       0.0
       ROUND_NUMBER  int64           0       0.0
         ROUND_PICK  int64           0       0.0
       OVERALL_PICK  int64           0       0.0
         DRAFT_TYPE object           0       0.0
            TEAM_ID  int64           0       0.0
          TEAM_CITY object           0       0.0
          TEAM_NAME object           0       0.0
  TEAM_ABBREVIATION object           0       0.0
       ORGANIZATION object           0       0.0
  ORGANIZATION_TYPE object           0       0.0
PLAYER_PROFILE_FLAG  int64           0       0.0


### `draft_history` — column reference

- **PERSON_ID** (`int64`) — unique player ID, matches `id` in `players_df`
- **PLAYER_NAME** (`object`) — player's full name at time of draft
- **SEASON** (`object`) — draft year
- **ROUND_NUMBER** (`int64`) — which round the player was picked in (1 or 2 in the modern draft; historically drafts had more rounds)
- **ROUND_PICK** (`int64`) — pick number within that round
- **OVERALL_PICK** (`int64`) — pick number overall across the whole draft
- **DRAFT_TYPE** (`object`) — type of draft event (e.g. standard "Draft")
- **TEAM_ID** (`int64`) — drafting team's ID, matches `id` in `teams_df`
- **TEAM_CITY** / **TEAM_NAME** / **TEAM_ABBREVIATION** (`object`) — drafting team's city, name, and 3-letter code at the time
- **ORGANIZATION** (`object`) — the college, international team, or organization the player came from
- **ORGANIZATION_TYPE** (`object`) — category of that organization (e.g. "College", "International")
- **PLAYER_PROFILE_FLAG** (`int64`) — flag indicating whether this player has a full profile page in the NBA's system (older/obscure picks may not)

Rows: 8,374 (every draft pick recorded, spanning nearly 80 years). Columns: 14. Nulls: 0 across all columns — draft records are administrative bookkeeping, complete regardless of era, unlike in-game stat tracking which has known historical gaps.

In [17]:
league_leaders = leagueleaders.LeagueLeaders(season="2025-26").get_data_frames()[0]
display(league_leaders.head())
print(f"rows: {league_leaders.shape[0]}, columns: {league_leaders.shape[1]}")

,PLAYER_ID,RANK,PLAYER,TEAM_ID,TEAM,GP,MIN,FGM,FGA,FG_PCT,...,REB,AST,STL,BLK,TOV,PF,PTS,EFF,AST_TOV,STL_TOV
0,1629029,1,Luka Dončić,1610612747,LAL,64,2289,693,1457,0.476,...,495,530,105,34,255,153,2143,2146,2.08,0.41
1,1628983,2,Shai Gilgeous-Alexander,1610612760,OKC,68,2259,731,1321,0.553,...,292,448,95,52,151,139,2117,2189,2.97,0.63
2,1627759,3,Jaylen Brown,1610612738,BOS,71,2443,736,1543,0.477,...,492,364,72,27,259,191,2038,1817,1.41,0.28
3,201142,4,Kevin Durant,1610612745,HOU,78,2840,716,1376,0.520,...,426,372,62,71,246,142,2026,1992,1.51,0.25
4,1630178,5,Tyrese Maxey,1610612755,PHI,70,2661,694,1501,0.462,...,290,461,130,55,171,151,1980,1893,2.70,0.76


rows: 582, columns: 28


In [18]:
league_leaders_summary = profile_dataframe(league_leaders, "league_leaders")

--- league_leaders ---
rows: 582, columns: 28
   column   dtype  null_count  null_pct
PLAYER_ID   int64           0       0.0
     RANK   int64           0       0.0
   PLAYER  object           0       0.0
  TEAM_ID   int64           0       0.0
     TEAM  object           0       0.0
       GP   int64           0       0.0
      MIN   int64           0       0.0
      FGM   int64           0       0.0
      FGA   int64           0       0.0
   FG_PCT float64           0       0.0
     FG3M   int64           0       0.0
     FG3A   int64           0       0.0
  FG3_PCT float64           0       0.0
      FTM   int64           0       0.0
      FTA   int64           0       0.0
   FT_PCT float64           0       0.0
     OREB   int64           0       0.0
     DREB   int64           0       0.0
      REB   int64           0       0.0
      AST   int64           0       0.0
      STL   int64           0       0.0
      BLK   int64           0       0.0
      TOV   int64           0     

### `league_leaders` — column reference

- **PLAYER_ID** (`int64`) — unique player ID, matches `id` in `players_df`
- **RANK** (`int64`) — player's overall rank for this leaderboard's default sort stat (points, by default)
- **PLAYER** (`object`) — player's full name
- **TEAM_ID** (`int64`) — team ID, matches `id` in `teams_df`
- **TEAM** (`object`) — 3-letter team abbreviation
- **GP** (`int64`) — games played
- **MIN** (`int64`) — total minutes played
- **FGM/FGA/FG_PCT**, **FG3M/FG3A/FG3_PCT**, **FTM/FTA/FT_PCT** — season shooting totals (made/attempted/percentage, for field goals, 3-pointers, free throws)
- **OREB/DREB/REB** (`int64`) — offensive, defensive, total rebounds
- **AST** (`int64`) — assists, **STL** — steals, **BLK** — blocks, **TOV** — turnovers, **PF** — personal fouls
- **PTS** (`int64`) — total points scored
- **EFF** (`int64`) — NBA efficiency rating, a single combined-value stat: `(PTS + REB + AST + STL + BLK) - ((FGA - FGM) + (FTA - FTM) + TOV)`
- **AST_TOV** (`float64`) — assist-to-turnover ratio (higher = more careful with the ball relative to playmaking)
- **STL_TOV** (`float64`) — steal-to-turnover ratio

Rows: 582 (every player with recorded stats this season). Columns: 28. Nulls: 0 across all columns — season-aggregated totals, same pattern as every other dashboard-style table in this notebook.

### ***Questions:***

- What's the importance of checking for nulls?
- Is collecting data from each endpoint the conventional way to pull from a dataset or is there a more optimized route?
- Everything I pulled had 0 on its null checks. What's another way to check if I'm doing the null checks correctly or the if the data is structered in a way where nulls are counted as 0?
- What are the pro and cons of grabbing static data, opposed to live endpoint data.
- Instead of relying on AI too much at times, how could I go about better understanding the syntax and logic?

In [1]:
print("test")

test
